quay_xfer.sh

```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+
+### requires jq + skopeo to be installed locally
+
+# arg check
+if [ $# -ne 2 ]; then
+  echo "Usage: $0 source_namespace/source_repo dest_namespace/dest_repo"
+  exit 1
+fi
+
+# extract the registry info
+SOURCE_REPO="$1"
+DESTINATION_REPO="$2"
+SOURCE_REGISTRY=$(echo "$SOURCE_REPO" | awk -F'/' '{print $1}')"
+DESTINATION_REGISTRY=$(echo "$DESTINATION_REPO" | awk -F'/' '{print $1}')"
+
+# check that
+if skopeo login --get-login "$SOURCE_REGISTRY" &> /dev/null && skopeo login --get-login "DESTINATION_REGISTRY" &> /dev/null; then
+    echo "Credentials for both repos authenticated"
+  else
+    echo "Error authenticating credentials"
+    echo "Login using skopeo login for both repos"
+    exit 1
+fi 
+
+# build skopeo command
+SKOPEO_CMD="skopeo copy --all"
+
+# fetch tags from the source repo with pagination
+TAGS=()
+page=1
+echo "Begin fetching tags"
+
+while :; do
```

Comment:
Others look good to me. I just don't understand why it uses `while :` here, endless loop, can anyone explan? Thanks!

Reply:
The JSON that's returned with [the curl call](https://github.com/konflux-ci/integration-service-utils/pull/6/files#diff-4e3a5863a11bddf49ea4793161750ca4f0a9dc0ba3f0e6b5ef6e341167523a43R36) is paginated but doesn't tell us how many pages there are.

Because of that need to cycle through pages one by one until the `has_additional` value is `false`, at which point we can [break out of the loop](https://github.com/konflux-ci/integration-service-utils/pull/6/files#diff-4e3a5863a11bddf49ea4793161750ca4f0a9dc0ba3f0e6b5ef6e341167523a43R56).

This while loop also has other break conditions such as when [running into an error](https://github.com/konflux-ci/integration-service-utils/pull/6/files#diff-4e3a5863a11bddf49ea4793161750ca4f0a9dc0ba3f0e6b5ef6e341167523a43R40) or [getting an empty/nonexistent page](https://github.com/konflux-ci/integration-service-utils/pull/6/files#diff-4e3a5863a11bddf49ea4793161750ca4f0a9dc0ba3f0e6b5ef6e341167523a43R47) so I think that provides enough safety catches to not run the script forever.

Here's an example from my testing repo to illustrate a JSON page that's returned with a single curl call:

```
{
  "tags": [
    {
      "name": "latest",
      "reversion": false,
      "start_ts": 1695806259,
      "manifest_digest": "sha256:2e4d8cc3de3b002a6f011990989a6ec250ad9993b9a27cfedac044509d98bc10",
      "is_manifest_list": false,
      "size": 263979037,
      "last_modified": "Wed, 27 Sep 2023 09:17:39 -0000"
    },
    {
      "name": "test",
      "reversion": false,
      "start_ts": 1653400564,
      "manifest_digest": "sha256:1efca91793e307ff48624de0f864e93acbeef52392c3e3e29a7d1d1cde23466b",
      "is_manifest_list": false,
      "size": 152519387,
      "last_modified": "Tue, 24 May 2022 13:56:04 -0000"
    },
    {
      "name": "gosec4",
      "reversion": false,
      "start_ts": 1650980169,
      "manifest_digest": "sha256:d53e99c8e1d036c8e20c7ebf645ea9ac7826e3458b481ff6c83d5e4a67930fc6",
      "is_manifest_list": false,
      "size": 307942954,
      "last_modified": "Tue, 26 Apr 2022 13:36:09 -0000"
    },
  ],
  "page": 1,
  "has_additional": false
}

```

Reply:
@dirgim Thanks very much for the details! I see now.